[MRG+1] IsolationForest max_samples warning and calculation

betatim opened the pull request:
In response to #5672

Introduces `max_samples='auto'` as the new default to make it possible to check if the user set a non default value. Only warn about `max_samples>n_samples` if the user set the value explicitly.

Fixes a bug where `max_depth` was not recalculated when `max_samples` has changed.

Now 256 appears as a magic value in both `fit` and `predict`, which isn't nice. I'll think about how to make it nicer, suggestions welcome.

todo:
- [x] add test that `max_depth` recalculation works
- [x] add a test to check the value of self.max_samples_ in both cases (=self.max_samples or =n_samples)?

/cc @ngoix

betatim committed:
Calculate max_depth when max_samples is known

max_samples='auto' replaces max_samples=256 as the default
setting. Recalculate max_depth based on the actual value
of max_samples after checking range and other constraints

ngoix commented on sklearn/ensemble/iforest.py:
```diff
@@ -151,15 +151,28 @@ def fit(self, X, y=None, sample_weight=None):
+                         % (self.max_samples, n_samples))
+                    max_samples = n_samples
+
+            else: # float
```
Here you are duplicating code from BaseBagging._fit(). You don't need L158,159,160 and paragraph L169.

ngoix commented on sklearn/ensemble/iforest.py:
```diff
@@ -47,10 +47,11 @@ class IsolationForest(BaseBagging):
         The number of samples to draw from X to train each base estimator.
             - If int, then draw `max_samples` samples.
             - If float, then draw `max_samples * X.shape[0]` samples.
+            - If "auto", then `max_samples=256`.
```
256 -> min(256, n_samples)

ngoix commented on sklearn/ensemble/iforest.py:
```diff
@@ -151,15 +151,28 @@ def fit(self, X, y=None, sample_weight=None):
+                else:
+                    raise ValueError('max_samples must be in (0, 1]')
+
+        self.base_estimator.max_depth = int(np.ceil(np.log2(max(max_samples, 2))))
```
I'm not sure if it is `self.base_estimator` or `self.base_estimator_` that you have to consider.

GaelVaroquaux replied:
Good catch. It should be self.base_estimator_

betatim replied:
We are trying to set a parameter of the instance passed as the `base_estimator` argument to `__init__` so I think it should be `self.base_estimator`

ngoix commented on sklearn/ensemble/iforest.py:
```diff
@@ -206,7 +219,9 @@ def predict(self, X):
 
         depths += _average_path_length(n_samples_leaf)
 
-        if not isinstance(self.max_samples, (numbers.Integral, np.integer)):
```
Here something is wrong (and was wrong before your changes): we should use the `max_samples` passed to the `base_estimator` (and in this case we don't need those 3 checks).
Use directly `max_samples = self.base_estimator.max_samples` ?

betatim replied:
The problem is that the `base_estimator` (`ExtraTreeRegressor` in this case) does not know anything about `max_samples`. It is just given a set of samples to fit to by `_parallel_build_estimators()`

betatim replied:
When is `max_samples_` set? The docs say it is the actual number of samples, which is what we are after but I can't find it being set anywhere?

ngoix replied:
Yes seems like we forgot to remove it from the doc, we didn't need it anymore. But we may need it now, since as you said base_estimator don't know `max_samples`...

ngoix replied:
And in that case you'll need to duplicate code as you've done before to set `max_samples_` in fit() I'm affraid.

betatim committed:
Fix docs and remove duplicated checks

betatim committed:
Comment on weird initialisation for scoping reasons

betatim committed:
Test for max_depth calculation and re-adding max_samples_ property

Added a test that checks if max_depth is recalculated for
small samples.

Readded the max_samples_ property to gain access to it in
predict()

betatim commented:
I re-added the `max_samples_` attribute to the classifier. This means we duplicate some code from `BaseBagging` in order to calculate the integer value for max_samples. Feels suboptimal but I can't see a better way. Do you remember why you initially decided against having the attribute? Maybe that reason is still valid and we should instead have `self._max_samples`?

Added a test to check that `max_depth` is calculated correctly in the case where `max_samples` is forced/limited to `n_samples`.

ngoix commented:
We introduced this attribute before adding the `_fit()` method in `BaseBagging`, which make it useless at that time.
This PR looks good to me, except that I'm still not 100% sure about L174 (`self.base_estimator` or `self.base_estimator_` ?), ping @agramfort. 
Could you just please add a test to check the value of `self.max_samples_` in both cases (=self.max_samples or =n_samples)?

GaelVaroquaux commented on sklearn/ensemble/iforest.py:
```diff
@@ -151,15 +151,27 @@ def fit(self, X, y=None, sample_weight=None):
+                max_samples = int(self.max_samples * X.shape[0])
+
+        self.max_samples_ = max_samples
+        self.base_estimator.max_depth = int(np.ceil(np.log2(max(max_samples, 2))))
```
It must be on base_estimator_, has you cannot modify an init argument at any time.

agramfort replied:
>
> It must be on base_estimator_, has you cannot modify an init argument at any time.
>
>
+1

betatim replied:
`base_estimator_` doesn't exist. There is a list of the fitted estimators (`self.estimators_`) but setting the property after fitting is too late. [BaseEnsemble](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/base.py#L50) explicitly delays instantiating the estimators so that the parameters of the `base_estimator` can still be changed. So I think this is what you want to be doing. The only alternative I see is not instantiating the base estimator until `fit()` is being called. For this we'd have to change `BaseEnsemble` though, as it expects you to pass a value for the `base_estimator` argument.

ngoix replied:
One other way would be to add an optional argument `max_depth` (argument to use instead of self.max_depth) to `_fit()` in `BaseBagging`, and to change `self.base_estimator_.max_depth` after `self._validate_estimator()` (L.290 in bagging.py). Indeed `_validate_estimator()` makes `base_estimator_` exist.

agramfort replied:
sounds reasonable to me.

betatim replied:
Sorry for being a bit thick, I think I get it now.

betatim committed:
Test max_samples_ attribute, set max_depth in BaseBagging

betatim commented:
Added a test for `max_samples_` being set correctly, as well as checking that `max_samples='illegal_string'` raises an exception.

Finally understood what you meant with `base_estimator` and `base_estimator_` and implemented @ngoix's proposal. Should we implement a test for the `max_depth` parameter? Do we need to check if it made sense to pass `max_depth` to `_fit` (eg when the base estimator isn't a tree) or is it assumed the user knows what they are doing because it is `_fit` and not `fit`?

agramfort commented:
design looks good !

you have a travis failure.

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -267,6 +267,9 @@ def _fit(self, X, y, max_samples, sample_weight=None):
             Argument to use instead of self.max_samples.
 
+        max_depth : int, optional (default=None)
+            Maximum depth of the base estimator.
```
you should say 
  ```
Argument to use instead of the one initially passed to the base estimator.
This is supported only if the base estimator has a max_depth parameter.
```

ngoix commented on sklearn/ensemble/iforest.py:
```diff
@@ -151,16 +151,29 @@ def fit(self, X, y=None, sample_weight=None):
+                else:
+                    max_samples = self.max_samples
+            else: # float
+                if not (0 < self.max_samples <= 1.):
```
0 -> 0.
This causes travis failure

ngoix commented:
Apart for this two minor comments, everything looks great to me. Thanks a lot @betatim :) !

betatim committed:
Fix doc string and number vs string ordering

betatim commented:
Turns out in `python` the ordering of numbers and strings isn't defined. Legacy python does define it though. So I extended the `isinstance` checking to check explicitly for `str` vs `int`.

Thanks for your patience! :cake:

agramfort commented:
@ngoix ok for you?

ngoix commented:
Yes everything looks fine. Any idea why appveyor is not happy?

giorgiop commented:
I don't why it only breaks for that particular instance (python 2.7.8 64bit), but the error is raised [here](https://github.com/betatim/scikit-learn/blob/no-warning-iforest/sklearn/ensemble/iforest.py#L275): 
`AttributeError: 'long' object has no attribute 'shape'`

giorgiop commented:
What happens if you change the else statement in `elif isinstance(n_samples_leaf, np.ndarray):`?

glouppe commented on sklearn/ensemble/bagging.py:
```diff
@@ -289,6 +293,9 @@ def _fit(self, X, y, max_samples, sample_weight=None):
+
         # if max_samples is float:
         if not isinstance(max_samples, (numbers.Integral, np.integer)):
             max_samples = int(self.max_samples * X.shape[0])
```
This is another left-out bug. It should be `max_samples = int(max_samples * X.shape[0])` (without `self`).

ngoix replied:
@betatim if you make this change appveyor should be happy.

glouppe commented on sklearn/ensemble/iforest.py:
```diff
@@ -151,16 +152,34 @@ def fit(self, X, y=None, sample_weight=None):
+        self.max_samples_ = max_samples
+        max_depth = int(np.ceil(np.log2(max(max_samples, 2))))
         super(IsolationForest, self)._fit(X, y, max_samples,
+                                          max_depth=max_depth,
```
I have been toying a bit with the algorithm and it seems results are really sensitive to `max_depth`. I would vote for making it public, with a default `"auto"` value set to `np.log2(max(max_samples, 2))`.

glouppe commented:
There is another thing that I dont currently understand in the codebase. In `predict`, scores are normalized by `_average_path_length(max_samples)`, why? In particular, the behaviour is different for float and integer values of `max_samples`. If float, then it becomes dependent on the size of `X` passed to predict, which means that scores will vary simply depending on the number of points you want to predict the abnormality score... This is obviously wrong. What do you think @ngoix ?

(Sorry @betatim for uncovering more bugs and make your PR longer :))

ngoix commented:
Yes you are right @glouppe it was wrong, but it is fixed now :) (see L229)

glouppe commented:
Oh yeah right, forget what I said :)

betatim committed:
Reference correct variable

betatim committed:
Use six to detect integral numbers

betatim commented:
Appveyor is unhappy because L266 checks if the argument is an `int` and if not assumes it is a `ndarray`. For appveyor we get passed a `long`, which isn't an `int`, and so end up in the `ndarray` branch.

Not quite sure where we get the `long` from in the first place. Should we track it down? This fix correctly detects `long`s and `int`s as integral numbers.

agramfort commented:
use six.integer_types to test if you have an int.

betatim commented:
Merci. The check I used was inspired by others in `ensemble/*` should I make an issue for this?

betatim commented:
All builds are green again.

agramfort commented:
+1 merge

@glouppe merge is you're happy too

glouppe commented:
Looking good, merging. Thanks Tim!

betatim commented:
:cake: 

Just noticed the number of this PR!
